kubevirt: add tests for Wrappers, EnhancedK8sMethods and utility classes

suomiy opened the pull request:
also fixes few bugs which were found by the testing

@vojtechszocs @glekner @irosenzw @mareklibra @yaacov @pcbailey

suomiy commented on frontend/packages/kubevirt-plugin/src/k8s/wrapper/common/__tests__/mocks/test-wrapper.ts:
```diff
@@ -0,0 +1,12 @@
+import { Wrapper } from '../../wrapper';
```
this file is a simple example of a `Wrapper`

suomiy commented on frontend/packages/kubevirt-plugin/src/k8s/wrapper/common/__tests__/mocks/smoothie-wrapper.ts:
```diff
@@ -0,0 +1,85 @@
+import { ObjectEnum } from '../../../../../constants';
```
this file is a simple example of `ObjectWithTypePropertyWrapper`

suomiy commented:
/retest

suomiy commented:
/retest

glekner commented on frontend/packages/kubevirt-plugin/src/k8s/wrapper/vm/vm-wrapper.ts:
```diff
@@ -176,35 +175,19 @@ export class VMWrapper extends K8sResourceWrapper<VMKind, VMWrapper> implements
-    if (_.isEmpty(this.getDataVolumeTemplates())) {
-      delete this.data.spec.dataVolumeTemplates;
-    }
+    this.clearIfEmpty('spec.template.spec.domain.devices.disks');
```
can't we merge `ensureStorages` and `ensureStorageConsistency` into one method?

suomiy replied:
no, one is called before the change of data occurs and the other is called after the change of data - please see `prependStorage` method

glekner commented on frontend/packages/kubevirt-plugin/src/k8s/enhancedK8sMethods/__tests__/enhancedK8sMethods.spec.ts:
```diff
@@ -0,0 +1,205 @@
+    await methods.k8sPatch(VirtualMachineModel, testVM, [testPatch]);
+    const actualState = methods.getActualState();
+
+    expect(actualState).toHaveLength(2);
```
can you destruct `actualState` for readability? or add a comment? I have no idea whats inside.

suomiy replied:
added a better comment to the `getActualState` method

suomiy replied:
and returns ` K8sResourceKind[]`

glekner commented on frontend/packages/kubevirt-plugin/src/k8s/enhancedK8sMethods/__tests__/enhancedK8sMethods.spec.ts:
```diff
@@ -0,0 +1,205 @@
+    expect(methodsUnchecked.history).toHaveLength(0);
+
+    await methods.k8sCreate(VirtualMachineModel, testVM);
+    expect(methods.getHistory()).toHaveLength(1);
```
can you swap the expects? so first we check that there is history and then we actually test that there isn't

suomiy replied:
the other way is better - otherwise it could happend that some object overwrites the first one we checked and we wouldn't know a difference

glekner commented on frontend/packages/kubevirt-plugin/src/k8s/enhancedK8sMethods/__tests__/enhancedK8sMethods.spec.ts:
```diff
@@ -0,0 +1,205 @@
+    const results = await methods.rollback();
+
+    expect(results).toHaveLength(2);
+    expect(methods.getActualState()).toHaveLength(0);
```
why is `getActualState()` empty? just to understand..

suomiy replied:
because the rollback worked and there are no living objects we created before on the cluster

glekner commented on frontend/packages/kubevirt-plugin/src/k8s/wrapper/common/__tests__/mocks/smoothie-wrapper.ts:
```diff
@@ -0,0 +1,85 @@
+      blueberry?: {
+        color?: string;
+      };
+      strawberry?: {
```
where is 🍍 ? 
😅

suomiy replied:
>.<~

glekner commented on frontend/packages/kubevirt-plugin/src/k8s/wrapper/common/__tests__/object-with-type-property-wrapper.spec.ts:
```diff
@@ -0,0 +1,252 @@
+    expect(emptySmoothie.hasType()).toBe(false);
+    expect(emptySmoothie.asResource()).toEqual({});
+  });
+  //
```
empty comment

suomiy replied:
fixed

glekner commented on frontend/packages/kubevirt-plugin/src/k8s/wrapper/common/__tests__/object-with-type-property-wrapper.spec.ts:
```diff
@@ -0,0 +1,252 @@
+    });
+  });
+
+  it('stores inconsistent data', () => {
```
I don't understand this test, can you write more informative comments? thanks

suomiy replied:
added a comment few lines bellow

glekner commented on frontend/packages/kubevirt-plugin/src/k8s/wrapper/common/k8s-resource-wrapper.ts:
```diff
@@ -4,7 +4,7 @@ import { apiVersionForModel, K8sKind, K8sResourceKind } from '@console/internal/
 import { K8sResourceKindMethods } from '../types/types';
 
-export class K8sResourceWrapper<
+export abstract class K8sResourceWrapper<
```
are we sure there is no use case of creating an instance of `K8sResourceWrapper`?

suomiy replied:
It is better if the user creates its own definition if needed. Now it also requires model to be supplied.

glekner commented on frontend/packages/kubevirt-plugin/src/utils/common.ts:
```diff
@@ -2,7 +2,16 @@ import * as _ from 'lodash';
 
 export const omitEmpty = (obj, justUndefined = false) => {
```
isn't there a `lodash` alternative so we can remove this code?

suomiy replied:
have not found one

vojtechszocs replied:
[`_.compact(array)`](https://lodash.com/docs/4.17.15#compact)

> Creates an array with all falsey values removed. The values `false`, `null`, `0`, `""`, `undefined`, and `NaN` are falsey.

vojtechszocs replied:
Alternatively, you could use `_.remove` or `_.without`.

suomiy replied:
Cannot be used very well. It should work on objects and arrays recursively with an option to switch between null x undefined

suomiy commented:
added few changes to the wrappers since last time
- each K8sResource kind stores its model
- deprecated `initializeFromSimpleData` in favor of simple instance init. The uses of initializeFromSimpleData still need to be removed in the followup as there are many instances of this.  @vojtechszocs 
- added a simpler method for creating wrappers into EnhancedK8sMethods

suomiy commented:
also added mixins for our two diverged wrappers - just the more complicated methods for now

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/k8s/enhancedK8sMethods/__tests__/enhancedK8sMethods.spec.ts:
```diff
@@ -0,0 +1,205 @@
+    })
+    .build();
+
+  it('records history and shows actualState', async () => {
```
```suggestion
  it('records history', async () => {
```

suomiy replied:
changed

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/k8s/enhancedK8sMethods/__tests__/enhancedK8sMethods.spec.ts:
```diff
@@ -0,0 +1,205 @@
+  });
+};
+
+describe('enhancedK8sMethods.js', () => {
```
```suggestion
describe('enhancedK8sMethods', () => {
```

I'd use the logical module name to describe the test suite.

suomiy replied:
done

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/k8s/enhancedK8sMethods/__tests__/enhancedK8sMethods.spec.ts:
```diff
@@ -0,0 +1,205 @@
+    expect.assertions(4);
+    let rollbackError;
+    try {
+      await methods.rollback();
```
We could use the [`toThrowError`](https://jestjs.io/docs/en/expect#tothrowerror) matcher here.

```ts
expect(methods.rollback).toThrowError('rollback');
// alternatively perform equality check against provided object
expect(methods.rollback).toThrowError(new Error('rollback'));
```

suomiy replied:
This does not work very well as the error must implement Constructable

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/k8s/enhancedK8sMethods/__tests__/enhancedK8sMethods.spec.ts:
```diff
@@ -0,0 +1,205 @@
+      rollbackError = e;
+    }
+    expect(rollbackError?.message).toEqual('rollback');
+    expect(rollbackError?.errors).toHaveLength(2);
```
`errors` seems like a non-standard [`Error`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error) property, what does it contain?

suomiy replied:
Added a comparison of the expected errors

suomiy replied:
contains an errors of objects which failed to cleanup during the rollback

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/k8s/wrapper/common/__tests__/object-with-type-property-wrapper.spec.ts:
```diff
@@ -0,0 +1,252 @@
+    });
+  });
+
+  it('merges types correctly', () => {
```
You could also say.. smoothly :smiley:

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/k8s/wrapper/common/object-with-type-property-wrapper.ts:
```diff
@@ -22,6 +22,9 @@ export abstract class ObjectWithTypePropertyWrapper<
     super(data, copy);
     this.TypeClass = typeClass;
     this.typeDataPath = typeDataPath;
+    if (!_.isFunction(typeClass.getAll) || typeClass.getAll === ObjectEnum.getAll) {
```
Wait, isn't `getAll` in `ObjectEnum` marked as `abstract` ?

suomiy replied:
Because it is a static method.

suomiy replied:
cannot be made abstract

vojtechszocs replied:
Oh, right.

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/k8s/wrapper/common/util/k8s-mixin.ts:
```diff
@@ -0,0 +1,47 @@
+  labels?: { [k: string]: string };
+  annotations?: { [k: string]: string };
+};
+export const initK8sObject = (
```
Nit: newline above.

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/k8s/wrapper/utils/__tests__/utils.spec.ts:
```diff
@@ -0,0 +1,55 @@
+    expect(testData).toEqual(before);
+  });
+
+  it('ensures path  with array', () => {
```
```suggestion
  it('ensures path with array', () => {
```

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/k8s/wrapper/utils/utils.ts:
```diff
@@ -1,6 +1,6 @@
 import * as _ from 'lodash';
 
-export const ensurePath = (data, path: string[] | string, value: any = {}) => {
+export const ensurePath = (data: object, path: string[] | string, value: any = {}) => {
```
```suggestion
export const ensurePath = (data: {}, path: string[] | string, value: any = {}) => {
```

suomiy commented on frontend/packages/kubevirt-plugin/src/k8s/enhancedK8sMethods/__tests__/enhancedK8sMethods.spec.ts:
```diff
@@ -0,0 +1,205 @@
+    methodsUnchecked.registerKind(kind);
+    methodsUnchecked.appendHistory(new HistoryItem(HistoryType.DELETE, data), enhancedOpts);
+  };
+  return { methods, methodsUnchecked };
```
made a different method in the spyEnhancedK8sMethods
methods.mockK8sGet to avoid the type incompatibilities 

using just methods outside of this one and using any typecasts

suomiy commented on frontend/packages/kubevirt-plugin/src/k8s/enhancedK8sMethods/__tests__/enhancedK8sMethods.spec.ts:
```diff
@@ -0,0 +1,205 @@
+const expectHistory = (history, expectedHistory) => {
+  expect(history).toHaveLength(expectedHistory.length);
+
+  expectedHistory.forEach((expectHistoryItem, idx) => {
```
you are right - this is much easier to read. It was meant to print an index which failed, but I guess it is not that useful.

suomiy commented:
fixed

suomiy commented:
/retest

suomiy commented:
added deprecated statements to all initializeFromSimpleData functions

vojtechszocs commented:
/lgtm
